index.src.html

```diff
@@ -596,6 +628,67 @@ <h3 id="monkeypatch-fetch">Fetch</h3>
+
+<h4 id="monkeypatch-html-navigation">Navigation</h4>
+
+During [=navigate|navigation=], in particular in the <a spec="HTML">create and initialize a <code>Document</code> object</a> algorithm, we need to set the newly-created {{Window}}'s [=WindowOrWorkerGlobalScope/origin policy IDs=]. To do so, as the last substep of the "Otherwise" step that creates a new realm and Window, modify the existing last step to save the created [=environment settings object=] as <var ignore>settingsObject</var>, and then add the following step:
+
+1. Set <var ignore>settingsObject</var>'s [=environment settings object/global object=]'s [=WindowOrWorkerGlobalScope/origin policy IDs=] to the result of [=create a frozen array|creating a frozen array=] from <var ignore>response</var>'s [=response/origin policy=]'s [=origin policy/IDs=].
+
+<h4 id="monkeypatch-html-worker">Running a worker</h4>
+
+In the <a spec="HTML">run a worker</a> algorithm, we need to set the newly-created {{WorkerGlobalScope}}'s [=WindowOrWorkerGlobalScope/origin policy IDs=]. To do so, inside the <a spec="HTML">perform the fetch</a> steps, after step 2 (which calls [=fetch=]) insert the following step:
+
+1. Set <var ignore>worker global scope</var>'s [=WindowOrWorkerGlobalScope/origin policy IDs=] to the result of [=create a frozen array|creating a frozen array=] from <var ignore>response</var>'s [=response/origin policy=]'s [=origin policy/IDs=].
+
+<h3 id="monkeypatch-sw">Service Worker</h3>
+
+Various parts of the Service Worker specification need patching to support the {{WindowOrWorkerGlobalScope/originPolicyIds}} feature.
+
+<h4 id="monkeypatch-sw-model">Model</h4>
+
+Every [=/service worker=] gains an associated <dfn for="service worker">origin policy IDs</dfn>, which is a <code><a interface>FrozenArray</a>&lt;<a interface>DOMString</a>></code>.
+
+<h4 id="monkeypatch-sw-update">Update algorithm</h4>
+
+The <a spec="SERVICE-WORKERS">update</a> algorithm needs to save the [=/origin policy=] [=origin policy/IDs=] from the [=response=] onto the [=/service worker=]. To do so, insert the following step after step 1:
+
+1. Let <var ignore>origin policy IDs</var> be the result of [=create a frozen array|creating a frozen array=] from <var ignore>response</var>'s [=response/origin policy=]'s [=origin policy/IDs=].
+
+and then the following step after step 15:
+
+1. Set <var ignore>service worker</var>'s [=service worker/origin policy IDs=] to <var ignore>origin policy IDs</var>.
+
+<h4 id="monkeypatch-sw-run">Run service worker algorithm</h4>
+
+The <a spec="SERVICE-WORKERS">run service worker</a> algorithm needs to copy the [=service worker/origin policy IDs=] from the [=/service worker=] onto the {{ServiceWorkerGlobalScope}}. To do so, insert the following step after step 7.7:
+
+1. Set <var ignore>workerGlobalScope</var>'s [=WindowOrWorkerGlobalScope/origin policy IDs=] to <var ignore>serviceWorker</var>'s [=service worker/origin policy IDs=].
```

Comment:
Is there a realm issue here? The frozen array is reused between realms.

Reply:
Great catch, thank you!